metricflow/query/query_parser.py

```diff
@@ -643,59 +679,90 @@ def _parse_metric_names(
+                )
+                structured_names.append(structured_name)
 
         dimension_specs = []
         time_dimension_specs = []
         partial_time_dimension_specs = []
         entity_specs = []
 
-        for qualified_name in qualified_linkable_names:
-            structured_name = StructuredLinkableSpecName.from_name(qualified_name)
+        for structured_name in structured_names:
             element_name = structured_name.element_name
             entity_links = tuple(EntityReference(element_name=x) for x in structured_name.entity_link_names)
             # Create the spec based on the type of element referenced.
             if TimeDimensionReference(element_name=element_name) in self._known_time_dimension_element_references:
-                if structured_name.time_granularity:
+                if structured_name.time_granularity and not structured_name.date_part:
                     time_dimension_specs.append(
                         TimeDimensionSpec(
                             element_name=element_name,
                             entity_links=entity_links,
                             time_granularity=structured_name.time_granularity,
                         )
                     )
+                # If date part is passed, remove requested granularity (to be overridden with default).
                 else:
                     partial_time_dimension_specs.append(
                         PartialTimeDimensionSpec(
                             element_name=element_name,
                             entity_links=entity_links,
+                            date_part=structured_name.date_part,
                         )
                     )
             elif DimensionReference(element_name=element_name) in self._known_dimension_element_references:
                 dimension_specs.append(DimensionSpec(element_name=element_name, entity_links=entity_links))
             elif EntityReference(element_name=element_name) in self._known_entity_element_references:
                 entity_specs.append(EntitySpec(element_name=element_name, entity_links=entity_links))
             else:
+                valid_group_bys_for_metrics = self._metric_lookup.element_specs_for_metrics(list(metric_references))
                 valid_group_by_names_for_metrics = sorted(
-                    x.qualified_name for x in self._metric_lookup.element_specs_for_metrics(list(metric_references))
+                    list(
+                        set(
+                            x.qualified_name if qualified_linkable_names else x.element_name
+                            for x in valid_group_bys_for_metrics
+                        )
+                    )
                 )
 
+                # If requested by name, show qualified name. If requested as object, show element name.
+                display_name = structured_name.qualified_name if qualified_linkable_names else element_name
```

Comment:
Oh, is this the same as above? How does the output differ? Is this fixing that thing where our suggested names just don't work correctly?

If so, we might want to factor this out into a function so we can put a docblock around it.

Reply:
This shouldn't change the previous behavior, so nope not a bug fix!
All the refactors to this function are to enable passing object `group_by` into `parse_and_validate_query`. Previously, at the start of `parse_and_validate_query`, we would parse the `group_by` objects into `qualified_names` and stop dealing with objects at all. That doesn't work with `date_part` since we can't parse `date_part` into a qualified name and still retain the granularity, so I refactored these functions to handle the full objects (figuring that is the direction we want to move in anyway).

Reply:
Sounds good, will do!